src/peft/tuners/oft/layer.py

```diff
@@ -14,106 +14,293 @@
+            return
+
+        warnings.warn("Scaling operation for OFT not supported! Automatically set scale to 1.")
+
+    def scale_layer(self, scale: float) -> None:
+        if scale == 1:
+            return
 
-    def reset_adapter_parameters(self, adapter_name: str):
-        nn.init.zeros_(self.oft_r[adapter_name])
+        for active_adapter in self.active_adapters:
+            if active_adapter not in self.oft_r.keys():
+                continue
 
-    def reset_adapter_parameters_random(self, adapter_name: str):
-        nn.init.kaiming_uniform_(self.oft_r[adapter_name], a=math.sqrt(5))
+            warnings.warn("Scaling operation for OFT not supported! Automatically set scale to 1.")
+
+    def unscale_layer(self, scale=None) -> None:
+        for active_adapter in self.active_adapters:
+            if active_adapter not in self.oft_r.keys():
+                continue
+
+            warnings.warn("Unscaling operation for OFT not supported! Keeping scale to 1.")
 
     def update_layer(
-        self,
-        adapter_name: str,
-        r: int,
-        module_dropout: float,
-        init_weights: bool,
-        coft: bool = False,
-        eps: float = 6e-5,
-        block_share: bool = False,
-        **kwargs,
-    ) -> None:
+        self, adapter_name, r, oft_block_size, module_dropout, coft, eps, block_share, init_weights
+    ):
+        """
+        Update the linear layer with trainable OFT weights. Override for other layer types.
+        """
         """Internal function to create oft adapter
 
         Args:
             adapter_name (`str`): Name for the adapter to add.
             r (`int`): Rank for the added adapter.
-            module_dropout (`float`): The dropout probability for disabling adapter during training.
-            init_weights (`bool`): Whether to initialize weights.
+            oft_block_size (`int`): The block size for added adapter.
+            module_dropout (`float`): The multiplicative dropout probability for disabling adapter blocks during training.
             coft (`bool`): Whether to use the constrained variant of OFT or not.
             eps (`float`):
                 The control strength of COFT. The freedom of rotation. Only has an effect if `coft` is set to True.
             block_share (`bool`): Whether to share the OFT parameters between blocks or not.
+            init_weights (`bool`): Whether to initialize weights.
         """
-        if r <= 0:
-            raise ValueError(f"`r` should be a positive integer value but the value passed is {r}")
+        # Initialize the MultiplicativeDropoutLayer for module_dropout > 0.0.
+        if module_dropout > 0.0:
+            oft_dropout_layer = MultiplicativeDropoutLayer(p=module_dropout)
+        else:
+            oft_dropout_layer = nn.Identity()
+        self.oft_dropout.update(nn.ModuleDict({adapter_name: oft_dropout_layer}))
+
+        if r == 0 and oft_block_size != 0:
```

Comment:
Can we move all these checks into the `OftConfig.__post_init__` instead? We want to find these errors as early as possible.

Reply:
Yes. Updated.